RELEASE/scripts/autoscend.ash

```diff
@@ -5258,42 +5254,67 @@ boolean LX_loggingHatchet()
 
 boolean LX_getStarKey()
 {
-	if (internalQuestStatus("questL10Garbage") < 9 || internalQuestStatus("questL11Shen") < 7)
+	//if we don't want the star key then skip this function
+	if(!get_property("auto_getStarKey").to_boolean())
 	{
 		return false;
 	}
-	if(!get_property("auto_getStarKey").to_boolean())
+
+	//if already have the star key or already used it, then stop running this function.
+	if(!needStarKey())
+	{
+		set_property("auto_getStarKey", false);
+		return false;
+	}
+	
+	//if we did not progress enough in giant garbage quest to reach hole in the sky then skip this function
+	//if shen quest is not far ahead enough also skip. because shen might send us there and it would be a waste of adventures if we already got the star key before he sends us there.
+	if (internalQuestStatus("questL10Garbage") < 9 || internalQuestStatus("questL11Shen") < 7)
 	{
 		return false;
 	}
+	
+	//can we reach the hole in the sky. need rocketship, except in kingdom of exploathing.
 	if(item_amount($item[Steam-Powered Model Rocketship]) == 0 && !in_koe())
 	{
 		return false;
 	}
-	if(!needStarKey())
+	
+	//pull a star chart when all following are true: not hardcore. at the door. no chart. no key. key not used.
+	if (!in_hardcore() && internalQuestStatus("questL13Final") == 5 && item_amount($item[Richard\'s Star Key]) == 0 && item_amount($item[Star Chart]) == 0 && !get_property("nsTowerDoorKeysUsed").contains_text("Richard's star key"))
+	{
+		pullXWhenHaveY($item[Star Chart], 1, 0);
+	}
+	
+	//craft the star key if I have the ingredients for it
+	if((item_amount($item[Richard\'s Star Key]) == 0) && (item_amount($item[Star Chart]) > 0) && (item_amount($item[star]) >= 8) && (item_amount($item[line]) >= 7) && !get_property("nsTowerDoorKeysUsed").contains_text("Richard's star key"))
+	{
+		visit_url("shop.php?pwd&whichshop=starchart&action=buyitem&quantity=1&whichrow=141");
+		if(item_amount($item[Richard\'s Star Key]) == 0)
+		{
+			cli_execute("make richard's star key");
+		}
+		return true;
+	}
+	
+	//wait to pull star chart at the door if: not hardcore, not at the door, have lines, have stars
+	if((item_amount($item[Star]) >= 8) && (item_amount($item[Line]) >= 7) && !in_hardcore() && internalQuestStatus("questL13Final") != 5)
 	{
-		set_property("auto_getStarKey", false);
 		return false;
 	}
+	
 	if(!zone_isAvailable($location[The Hole In The Sky]))
 	{
 		auto_log_warning("The Hole In The Sky is not available, we have to do something else...", "red");
 		return false;
 	}
 
-	if((item_amount($item[Star]) >= 8) && (item_amount($item[Line]) >= 7))
-	{
-		if(!in_hardcore())
-		{
-			set_property("auto_getStarKey", false);
-			return false;
-		}
-	}
-	else
+	//if you don't have space jellyfish get an item boosting familiar. If you do then get the space jellyfish
+	if(!auto_have_familiar($familiar[Space Jellyfish]))
```

Comment:
nit: please invert to `if(auto_have_familiar($familiar[space jellyfish])`

why`if (!a) {} else {}`

Reply:
change applied on commit d1b54fb